src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java

```diff
@@ -0,0 +1,153 @@
+            // doing it yourself when necessary.
+            if (transaction instanceof Transaction resolvedTransaction)
+                resolvedTransaction.addClosingJournal(this);
+        }
+    }
+
+    /**
+     * Perform an action when a transaction is closed.
+     *
+     * @param transaction The closed transaction. Only {@link Transaction#depth()}, {@link Transaction#lifecycle()}, {@link Transaction#addCommittingJournal(SnapshotJournal)},
+     *                    {@link Transaction#addClosingJournal(SnapshotJournal)}, and {@link Transaction#addClosingJournalToPrevDepth(SnapshotJournal)} may be called on that transaction.
+     * @param wasAborted  {@code true} if the transaction was aborted, {@code false} if it should be committed.
+     */
+    @ApiStatus.OverrideOnly
+    protected void onClose(Transaction transaction, boolean wasAborted) {
+        int currentDepth = transaction.depth();
+        // For testing and will be removed after deprecation period is over for handler reworks.
+        // This is to provide a quick way to give some metrics during the migration phase, unless another route can be decided on.
+        SnapshotJournalDebugInfo.updateDeepestSnapshot(currentDepth, this);
+
+        // Get and remove the relevant snapshot.
+        T snapshot = snapshots.remove(currentDepth);
+
+        if (wasAborted) {
+            // If the transaction was aborted, we just revert to the state of the snapshot.
+            revertToSnapshot(snapshot);
+            releaseSnapshot(snapshot);
+        } else if (currentDepth <= 0) {
+            // The transaction is the root.
+            originalState = snapshot;
+            transaction.addCommittingJournal(this);
+        } else if (snapshots.get(currentDepth - 1) == null) {
+            // No snapshot yet, so move the snapshot one depth up.
+            snapshots.set(currentDepth - 1, snapshot);
+            // This is the first snapshot at this level: we need to add the closing journal to the previous depth.
+            transaction.addClosingJournalToPrevDepth(this);
+        } else {
+            // There is already an older snapshot at the depth above, just release the newer one.
+            releaseSnapshot(snapshot);
+        }
+    }
+
+    /**
+     * Called after the root transaction succeeded,
+     * to perform irreversible actions such as {@code setChanged()} or neighbor updates.
+     *
+     * @param originalState state of this journal before the transactional operation.
+     *                      This corresponds to the first {@link #createSnapshot() snapshot} that was created in the transactional operation.
+     * @throws IllegalStateException when trying to open a new transaction during this method as the current transaction is still in the process of closing.
+     */
+    @ApiStatus.OverrideOnly
+    protected void onRootCommit(T originalState) {}
```

Comment:
Since this method is very often overridden, I would move it just after `releaseSnapshot` so that it comes before the huge `onClose`.